### ISSUE TITLE
Fiber::yield called in internal callback cause core dump

### ISSUE THREAD

**Issue by lvht** [fiberphp/fiber-ext#2]
The following code will cause a core dump.

```php
$f = new Fiber(function () {
    array_map(function ($i) {
        Fiber::yield($i);
    }, [1,2]);
});

$f->resume();
```

**Comment by kelunik** [comment]
Please re-open, so remediation plans can be discussed. `array_map` is rather easy to avoid, but PHPUnit uses reflection to find the tests it needs to execute, which means no Fiber-using code can be tested with PHPUnit.

**Comment by bwoebi** [comment]
Yeah, for using this in production for real (and inclusion within php-src), supporting this is a must.

(via makecontext/setcontext/getcontext/swapcontext functions)

**Comment by kelunik** [comment]
@bwoebi `getcontext` has been dropped in POSIX 2008.

Here's a possible implementation: https://github.com/brianwatling/libfiber/blob/bcb527c2ae5d3e66be8867061196d6cce7f00321/src/fiber_context.c#L160-L256

**Comment by kelunik** [comment]
I guess the implementation of `fiber_context.c` / `fiber_context.h` / `machine_specific.h` could be copied (with appropriate attribution of course) and integrated.

**Comment by joelwurtz** [comment]
Isn't this only a problem on the fiber not being terminated and thus the callback cannot be cleared problem ?

The following code works fine : 


```

<?php
$f = new Fiber(function () {
    return array_map(function ($i) {
        return Fiber::yield($i);
    }, [1,2]);
});

var_dump($f->resume());
var_dump($f->resume(3));
var_dump($f->resume(4));
```

Can add the test if you want

**Comment by trowski** [comment]
Interestingly, this also works:

```php
$f = new Fiber(function () {
    return array_map(function ($i) {
        return Fiber::yield($i);
    }, ['a', 'b']);
});


$f2 = new Fiber(function () {
    return array_map(function ($i) {
        return Fiber::yield($i);
    }, [1,2]);
});

var_dump($f->resume());
var_dump($f->resume(3));
var_dump($f->resume(4));

var_dump($f2->resume());
var_dump($f2->resume(4));
var_dump($f2->resume(6));
```

Could this be as simple as keeping track of some state in the VM and restoring it accordingly if switching between different fibers when in an internal call?

**Comment by bwoebi** [comment]
No. Also @trowski this only works because array_map() is only started after the initial `Fiber->resume()` call. But once we start having two internal functions in two different active fibers at once, it will be all over.

**Comment by trowski** [comment]
@bwoebi Something that we can accomplish with modifying the PHP VM, or do we have to resort to C stack switching?

**Comment by bwoebi** [comment]
Not that I'm aware of. Unless we're moving internal fcalls into the VM in PHP (i.e. `array_map($x, $array)` would need to return to vm for each function call and have the VM call back into the C function upon each termination of the function). This still doesn't allow for __toString(), but we're already anyway forbidding exceptions being thrown from it (for the same reason).

Basically we need to fix every use of zend_call_function and zend_call_method (about 35 occurrences).

It'd be a quite a bit more invasive patch, but could be worth it though. Especially if we want to integrate this into php-src.

**Comment by kelunik** [comment]
I think moving those into the VM is the better solution, as we don't have to care about different CPU architectures for doing the context switch then.

**Comment by nikic** [comment]
I'm pretty sure that moving this into the VM is not going to happen. That's something that might work for array_map, but not for all invocations of userland callbacks in all extensions (which can involve stuff like, say PHP calling into SQLite and SQLite then calling back into PHP, etc.)

I think the simplest way to implement this for now is to create a new thread for each fiber. That may not be the most efficient way, but easier than mucking about with platform specific stack context switching.

**Comment by uasan** [comment]
> create a new thread for each fiber

Does this mean that `Fiber` will work only on the PHP ZTS (Zend Thread Safety)?

**Comment by bwoebi** [comment]
ZTS is unrelated and is about allowing multiple (independent, in case of pthreads: dependent) PHP scripts in _parallel_.
If we were to use a thread per fiber (a good idea for initial support, before we move that into the VM as a possible future optimization), the threads would only ever run sequentially (on the high level every thread has a lock except one. If control is yielded to another fiber, it releases the lock of another thread and locks itself.) and not running into any races.

**Comment by trowski** [comment]
Threads sounds like a reasonable compromise then, especially if we could pool idle threads.

**Comment by kelunik** [comment]
But I guess kernel level threads are too expensive there if you have lots of idle fibers.

**Comment by trowski** [comment]
@kelunik I was thinking we could keep `Fiber` objects in a pool and create and destroy them as needed. I noticed @lvht removed `Fiber::reset()` (not sure on why), but if this was restored fibers could then be reused or destroyed when idle for too long.

**Comment by kelunik** [comment]
That improves the situation, but I guess it will still be problematic if you have 1k or 10k open websocket connections.

**Comment by trowski** [comment]
Hmm… yes, having a thread-per-client isn't really practical, but each client would need it's own fiber. Looks like some other solution is needed then.

**Comment by trowski** [comment]
Could internal calls from within a fiber context be forbidden? That is, simply throwing an exception from `Fiber::yield()` if invoked within an internal call? In general most internal calls do not need to do I/O, so therefore do not need to be fibers. For the few that are useful (such as `array_map`), it's trivial to implement a fiber-aware and cooperative version in user-code.

**Comment by lvht** [comment]
@trowski 
> Could internal calls from within a fiber context be forbidden?

This is also my solution to solve this problem.

And here is a simple idea. @nikic 

Every time we init a new vm instance to do execute user land php code, we increment a `global c frame counter`. When the execution finished, we decrement the `global c frame counter`. When we create a fiber, we save the current `global c frame counter` and when we call `Fiber::yield`, we should check wether the current `global c frame counter` is equal to the initial one.

However, every time you call the `Fiber::yield`, you need to do the additional check. As @trowski said
> In general most internal calls do not need to do I/O, so therefore do not need to be fibers. For the few that are useful (such as array_map), it's trivial to implement a fiber-aware and cooperative version in user-code.

I think the best way to solve this problem is let user know this problem and **DO NOT USE IT THEMSELVES**.

**Comment by uasan** [comment]
> so therefore do not need to be fibers. For the few that are useful (such as array_map), it's trivial to implement a fiber-aware and cooperative version in user-code.

It might be better to make another class `FiberIterator` that will be as an asynchronous iterator and be able to work with native `array_map()`, etc?

Something like [Asynchronous Iterators for JavaScript](https://github.com/tc39/proposal-async-iteration).

**Comment by lvht** [comment]
@uasan user land code

**Comment by uasan** [comment]
> by user land code

This does not work in userland, for the same reason:

```PHP
class FiberIterator implements Iterator
{
    public function next()
    {
        Fiber::yield(); // this problem
    }
}
```

But maybe it will be easier for you to implement in Fiber a special separate class that will work in functions such as array_map (), but I agree that this is not critical.

Thanks.

**Comment by trowski** [comment]
@uasan An approach like `while ($value = Fiber::yield($promise)) { ... }` works well in user-land. If someday keywords are introduced on top of fibers, then perhaps the changes necessary in the VM can be made to accommodate something like `foreach await ($asyncIterator as $key => value)` and async generators using `yield await $promise`.

**Comment by kelunik** [comment]
IMO we should fix the functions where it is possible and only error for the few where it is not. Everything else will be a mess. PHPUnit won't work for example. Also iterators won't work as pointed out, but they could work just fine. There's no need for that await there with fibers.

**Comment by martinschroeder** [comment]
I re-implemented fibers using native APIs to do the context switching. [The fork](https://github.com/martinschroeder/fiber-ext) uses inline Assembler (x86 and x86_64 ABI), `makecontext()` / `swapcontext()` or native Windows fibers depending on the OS. It works with PHP internal function calls, Generators and Iterators (see test cases).

A downside of this approach is that (in addition to a VM stack) a C stack has to be allocated for each fiber, which uses more memory than the VM interrupt-based version (it uses mmap if available to allocate virtual memory). Another downside is that the implementation is platform-specific and ABIs like ARM or MIPS are not supported yet.

The cool thing about this (re)implementation is that there is no need to make any changes to PHP builtin-functions or the VM allowing the extension to be tested and used with PHP 7.3+ without a patch or changes to the core.

I have tested the extension on Windows (7 & 10) and Ubuntu (x86_64 assembler and ucontext backend) and it worked just fine. Maybe someone could test it on Mac OS as i do not have that installed.

**Comment by kelunik** [comment]
@martinschroeder Awesome, will give that a try as soon as time allows!

**Comment by kelunik** [comment]
@martinschroeder Your implementation segfaults for me. Could you try running the tests of https://github.com/amphp/byte-stream/tree/ext-fiber?

**Comment by kelunik** [comment]
@martinschroeder Seems like it's exception related. https://github.com/amphp/byte-stream/commit/b7c91a2cb3aa06b6bfab2d53d57c2716b2c48860 and https://github.com/amphp/byte-stream/commit/3aa80afad8b66dbd30ed0c6d6046952b97e08a20 "fixed" two of the segfaults.

**Comment by kelunik** [comment]
@martinschroeder Could you open a PR so we can merge your patch?

**Comment by martinschroeder** [comment]
I create PR #30 and added some comments that outline a few of the code fragments that need extensive testing by someone with a better understanding of zvals and how PHP deals with them.

**Comment by need-sleep** [comment]
whats the state of this issue?

**Comment by martinschroeder** [comment]
@need-sleep I have stopped working on this extension and created [ext-async](https://github.com/concurrent-php/ext-async) instead. The new extension solves the problem with function calls from internal functions in the same way #30 did. It still provides fibers but using tasks is preferred (they come with a builtin scheduler).
